### ISSUE TITLE
.dockerignore does not recursively include with negated exclusion

### ISSUE THREAD

**Issue by snthibaud** [docker/cli#2847]
**Description**

The following .dockerignore file includes only .py files on the top-level, but not in sub-directories.

```
*
!**/*.py
!requirements.txt
```

Dockerfile includes:
```
COPY . .
```

**Steps to reproduce the issue:**
1. Build image
2. Run image and list files (I used Python's os.walkdir)

**Describe the results you received:**
Only .py files in root directory were included

**Describe the results you expected:**
.py files in subdirectories are included too

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Cloud integration: 1.0.2
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        4484c46d9d
 Built:             Wed Sep 16 16:58:31 2020
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.13
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       4484c46d9d
  Built:            Wed Sep 16 17:07:04 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

```

**Output of `docker info`:**

```
Client:
 Debug Mode: false
 Plugins:
  scan: Docker Scan (Docker Inc., v0.3.4)

Server:
 Containers: 230
  Running: 0
  Paused: 0
  Stopped: 230
 Images: 887
 Server Version: 19.03.13
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8fba4e9a7d01810a393d5d25a3621dc101981175
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.4.39-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 6
 Total Memory: 4.822GiB
 Name: docker-desktop
 ID: H3TG:YSVP:EPSW:MORJ:TSGF:ZNVG:KQS5:GP5I:ZCCI:5T2M:2CDH:O2XW
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 82
  Goroutines: 47
  System Time: 2020-11-19T10:50:26.2565802Z
  EventsListeners: 3
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Comment by adeniyistephen** [comment]
@thaJeztah I would love to work on this, Please provide pointers and input on how to work on it. Thanks

**Comment by thaJeztah** [comment]
this looks like a duplicate of https://github.com/moby/moby/issues/30018

I also created an "epic" issue for changes/fixes needed for `.dockerignore` in general; https://github.com/moby/moby/issues/40319; if you're interested in looking into that, that'd be welcome (it may not be an easy one to fix though, also w.r.t. making sure changes are backward-compatible if possible).

I should note that the "dockerignore" code now (since https://github.com/moby/buildkit/pull/1766) lives in the BuildKit repository, so changes for this need to be made in that repository.

Let me close this ticket to not have the discussion diverge as it's a duplicate of  https://github.com/moby/moby/issues/30018), but feel free to comment after I closed 😅
